src/gt4py/eve/utils.py

```diff
@@ -402,6 +402,48 @@ def hashable_by(
 hashable_by_id = hashable_by(id)
 
 
+def lru_cache(
+    func: Optional[Callable[_P, _T]] = None,
+    *,
+    key: Callable[_P, Any],
+    maxsize: Optional[int] = 128,
+    typed: bool = False,
+) -> Union[Callable[_P, _T], Callable[[Callable[_P, _T]], Callable[_P, _T]]]:
+    """
+    Wrap :func:`functools.lru_cache` but allow customizing the cache key.
+
+    Be careful: `key(obj1) == key(obj2)` must imply `obj1 == obj2`.
+
+    >>> @lru_cache(key=id)
+    ... def func(x):
+    ...     print("called")
+    ...     return x
+
+    >>> obj = object()
+    >>> func(obj) is obj
+    called
+    True
+    >>> func(obj) is obj
+    True
+    """
+
+    def _decorator(func: Callable[_P, _T]) -> Callable[_P, _T]:
+        @functools.lru_cache(maxsize=maxsize, typed=typed)
+        def cached_func(*args: HashableBy, **kwargs: HashableBy) -> _T:
+            return func(*(arg.value for arg in args), **{k: v.value for k, v in kwargs.items()})
+
+        @functools.wraps(func)
+        def inner(*args, **kwargs):  # type: ignore[no-untyped-def]  # cast below restores type info
+            return cached_func(
+                *(hashable_by(key, arg) for arg in args),
+                **{k: hashable_by(key, arg) for k, arg in kwargs.items()},
+            )
+
```

Comment:
I think this is also needed to honor the `cache_parameters()` instrumentation function (https://devdocs.io/python~3.13/library/functools#functools.lru_cache).
```suggestion
        inner.cache_parameters = cached_func.cache_parameters
```